### DISCUSSION TITLE
What is the recommended way to isolate test executions at a db level?

### DISCUSSION THREAD

**Discussion by sbalay** [prisma/prisma#6028] [Q&A]
Hi, I'm setting up a nodejs project using prisma and I'm trying to find a way to run my tests and guarantee that each test case will run with an empty db state.

I've read about [prisma-test-utils](https://github.com/prisma/prisma-test-utils) but seems to be only for prisma v1, and not for v2. ([this issue](https://github.com/prisma/prisma-test-utils/issues/6) mentions that at some point there will be a test utils lib for v2)

Also, I run into this setup example https://github.com/ctrlplusb/prisma-pg-jest. It's creates a new schema for each test suite run which is kinda slow, and also it doesn't guarantee that each test case will be isolated.

The best thing I've found so far is [this comment](https://github.com/prisma/prisma/issues/742#issuecomment-776901281) that truncates each table on the db, so Im calling that before each test but now running into some issues and not sure why.
I'm looking forward to https://github.com/prisma/prisma/issues/5596

Is there an official guide for testing? What do people usually set up their tests?

**Comment by ryands17** [comment]
@sbalay 👋 

I am not aware of unit testing, but you can perform integration tests as shown in the example [here](https://github.com/ryands17/fastify-prisma/blob/main/tests/auth.test.ts). This setup uses Postgres as the database and a REST API.

As for setting up and tearing down, have a look at [this](https://github.com/ryands17/fastify-prisma/blob/main/prisma/prisma-test-environment.js) file for more details.

**Reply by sbalay** [reply]
Thanks for the quick reply!

Does the setup/teardown config that you shared truncate the db between each test case? or only one time at the start of tests execution? For example:

```ts
describe('createUser', () => {
  it('test 1', async () => {
    const user = await prismaClient.user.create({
      data: {
        username: 'eeee',
      },
    })

    expect(user.id).toBe(1)
  })
  
  it('test 2', async () => {
    const usersCount = await prismaClient.user.count()
    expect(usersCount).toBe(0)
  })
})
```

Can I guarantee somehow that in `test 1` id will be `1` and that in `test 2` the count will be zero?

**Reply by ryands17** [reply]
This setup will ensure truncation between each test suite/file and not each individual test. So multiple test files running in parallel will have their own database.

**Reply by sbalay** [reply]
@ryands17 thanks for the clarification and help! I will probably go with this setup, but I think that it would be nice to have some way of quickly and easily cleaning the db between each test case.

[Here is an example from the ruby on rails community](https://github.com/DatabaseCleaner/database_cleaner#rspec-example), which I've seen applied in a lot of projects.

**Reply by sbalay** [reply]
Also, I haven't tried the proposed strategy yet, but recreating the schema each time before the tests may not be the fastest solution

**Reply by ryands17** [reply]
> Here is an example from the ruby on rails community, which I've seen applied in a lot of projects.

I don't think this solution is possible as Prisma handles transactions a [bit differently](https://www.prisma.io/blog/how-prisma-supports-transactions-x45s1d5l0ww1). So a special wrapper would need to be created for rolling back any operations done in the individual test block.

> but recreating the schema each time before the tests may not be the fastest solution

It's not re-creating the schema for each test, but uses multiple schemas so the tests can run in parallel and each test suite has its own schema, so speed shouldn't be much of an issue.

**Reply by sbalay** [reply]
> I don't think this solution is possible as Prisma handles transactions a bit differently. So a special wrapper would need to be created for rolling back any operations done in the individual test block.

But could the `truncation` strategy still be used? 

> It's not re-creating the schema for each test, but uses multiple schemas so the tests can run in parallel and each test suite has its own schema, so speed shouldn't be much of an issue.

Thanks again for the clarification! But we still run `await exec('yarn prisma db push --preview-feature')` on each schema, right?
Initially I thought that this runs each migration and that's why I assumed it could be slow. It just pushes the schema to the db, right?
I wonder if a truncation strategy could still be significantly faster, pretty much the same or even worse...

**Reply by ryands17** [reply]
> But could the truncation strategy still be used?

Yes it surely can, but you would need to write an `afterEach` block with the truncate logic :)

> I wonder if a truncation strategy could still be significantly faster, pretty much the same or even worse...

I personally haven't tried that but it would be a great point to explore 👍

**Reply by Jolg42** [reply]
Truncating should be faster yes.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
